# Attribute Extraction
October is close to over. In this notebook I explore pulling attributes from FMA dataset for my subset of 95 `.mp3`s I have extracted in the last notebook.

**Status so far:**
- FMA Dataset is back up
- 95 `.mp3` files downloaded

**Goals:**
- New dataframe has actual `mp3` *URLS*, but I'd like to add:
    - Local path
    - Attributes
- Clean up NULLs in new df
- Set up Google Colab once this data is fresh n ready
- Backburner: move out of notebooks eventually..?

In [5]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd

In [19]:
track_attr = pd.read_csv("track_attr.csv")
track_attr.head()

Unnamed: 0  track_id  album_id           album_title  \
0           0         2       1.0  AWOL - A Way Of Life   
1           1         3       1.0  AWOL - A Way Of Life   
2           2         5       1.0  AWOL - A Way Of Life   
3           3        10       6.0     Constant Hitmaker   
4           4        20       4.0                 Niris   

                                           album_url  artist_id artist_name  \
0  http://freemusicarchive.org/music/AWOL/AWOL_-_...          1        AWOL   
1  http://freemusicarchive.org/music/AWOL/AWOL_-_...          1        AWOL   
2  http://freemusicarchive.org/music/AWOL/AWOL_-_...          1        AWOL   
3  http://freemusicarchive.org/music/Kurt_Vile/Co...          6   Kurt Vile   
4  http://freemusicarchive.org/music/Chris_and_Ni...          4  Nicky Cook   

                                          artist_url  \
0            http://freemusicarchive.org/music/AWOL/   
1            http://freemusicarchive.org/music/AWOL/   
2            http://freemusicarchive.org/music/AWOL/   
3       http://freemusicarchive.org/music/Kurt_Vile/   
4  http://freemusicarchive.org/music/Chris_and_Ni...   

                            artist_website  \
0  http://www.AzillionRecords.blogspot.com   
1  http://www.AzillionRecords.blogspot.com   
2  http://www.AzillionRecords.blogspot.com   
3                      http://kurtvile.com   
4                                      NaN   

                                  license_image_file  ... track_instrumental  \
0  http://i.creativecommons.org/l/by-nc-sa/3.0/us...  ...                  0   
1  http://i.creativecommons.org/l/by-nc-sa/3.0/us...  ...                  0   
2  http://i.creativecommons.org/l/by-nc-sa/3.0/us...  ...                  0   
3  http://i.creativecommons.org/l/by-nc-nd/3.0/88...  ...                  0   
4  http://i.creativecommons.org/l/by-nc-nd/3.0/88...  ...                  0   

   track_interest track_language_code track_listens track_lyricist  \
0            4656                  en          1293            NaN   
1            1470                  en           514            NaN   
2            1933                  en          1151            NaN   
3           54881                  en         50135            NaN   
4             978                  en           361            NaN   

   track_number  track_publisher      track_title  \
0             3              NaN             Food   
1             4              NaN     Electric Ave   
2             6              NaN       This World   
3             1              NaN          Freeway   
4             3              NaN  Spiritual Level   

                                           track_url  \
0  http://freemusicarchive.org/music/AWOL/AWOL_-_...   
1  http://freemusicarchive.org/music/AWOL/AWOL_-_...   
2  http://freemusicarchive.org/music/AWOL/AWOL_-_...   
3  http://freemusicarchive.org/music/Kurt_Vile/Co...   
4  http://freemusicarchive.org/music/Chris_and_Ni...   

                                                urls  
0  https://files.freemusicarchive.org/storage-fre...  
1  https://files.freemusicarchive.org/storage-fre...  
2  https://files.freemusicarchive.org/storage-fre...  
3  https://files.freemusicarchive.org/storage-fre...  
4                                                  8  

[5 rows x 41 columns]

In [22]:
# temp until NULLs are polished
def is_dl(row):
    if track_attr["urls"].isnumeric():
        return False
    return True

# test:
track_attr["urls"][4].isnumeric()

True

In [24]:
feats = pd.read_csv("/Users/mkarroqe/Desktop/github/dancing-screen/fma_metadata/features.csv")
feats.head()

feature       chroma_cens     chroma_cens.1     chroma_cens.2  \
0  statistics          kurtosis          kurtosis          kurtosis   
1      number                01                02                03   
2    track_id               NaN               NaN               NaN   
3           2  7.1806526184e+00  5.2303090096e+00  2.4932080507e-01   
4           3  1.8889633417e+00  7.6053929329e-01  3.4529656172e-01   

      chroma_cens.3     chroma_cens.4     chroma_cens.5     chroma_cens.6  \
0          kurtosis          kurtosis          kurtosis          kurtosis   
1                04                05                06                07   
2               NaN               NaN               NaN               NaN   
3  1.3476201296e+00  1.4824777842e+00  5.3137123585e-01  1.4815930128e+00   
4  2.2952005863e+00  1.6540306807e+00  6.7592434585e-02  1.3668476343e+00   

      chroma_cens.7     chroma_cens.8  ...        tonnetz.39  \
0          kurtosis          kurtosis  ...               std   
1                08                09  ...                04   
2               NaN               NaN  ...               NaN   
3  2.6914546490e+00  8.6686819792e-01  ...  5.4125156254e-02   
4  1.0540937185e+00  1.0810308903e-01  ...  6.3831120729e-02   

         tonnetz.40        tonnetz.41               zcr             zcr.1  \
0               std               std          kurtosis               max   
1                05                06                01                01   
2               NaN               NaN               NaN               NaN   
3  1.2225749902e-02  1.2110591866e-02  5.7588901520e+00  4.5947265625e-01   
4  1.4211839065e-02  1.7740072682e-02  2.8246941566e+00  4.6630859375e-01   

              zcr.2             zcr.3             zcr.4             zcr.5  \
0              mean            median               min              skew   
1                01                01                01                01   
2               NaN               NaN               NaN               NaN   
3  8.5629448295e-02  7.1289062500e-02  0.0000000000e+00  2.0898721218e+00   
4  8.4578499198e-02  6.3964843750e-02  0.0000000000e+00  1.7167237997e+00   

              zcr.6  
0               std  
1                01  
2               NaN  
3  6.1448108405e-02  
4  6.9330163300e-02  

[5 rows x 519 columns]

In [25]:
feats.head(10)

feature        chroma_cens      chroma_cens.1      chroma_cens.2  \
0  statistics           kurtosis           kurtosis           kurtosis   
1      number                 01                 02                 03   
2    track_id                NaN                NaN                NaN   
3           2   7.1806526184e+00   5.2303090096e+00   2.4932080507e-01   
4           3   1.8889633417e+00   7.6053929329e-01   3.4529656172e-01   
5           5   5.2756297588e-01  -7.7654317021e-02  -2.7961030602e-01   
6          10   3.7022454739e+00  -2.9119303823e-01   2.1967420578e+00   
7          20  -1.9383698702e-01  -1.9852678478e-01   2.0154602826e-01   
8          26  -6.9953453541e-01  -6.8415790796e-01   4.8824872822e-02   
9          30  -7.2148716450e-01  -8.4855991602e-01   8.9090377092e-01   

       chroma_cens.3      chroma_cens.4      chroma_cens.5      chroma_cens.6  \
0           kurtosis           kurtosis           kurtosis           kurtosis   
1                 04                 05                 06                 07   
2                NaN                NaN                NaN                NaN   
3   1.3476201296e+00   1.4824777842e+00   5.3137123585e-01   1.4815930128e+00   
4   2.2952005863e+00   1.6540306807e+00   6.7592434585e-02   1.3668476343e+00   
5   6.8588310480e-01   1.9375696182e+00   8.8083887100e-01  -9.2319184542e-01   
6  -2.3444947600e-01   1.3673638105e+00   9.9841135740e-01   1.7706941366e+00   
7   2.5855624676e-01   7.7520370483e-01   8.4794059396e-02  -2.8929358721e-01   
8   4.2658798397e-02  -8.1896692514e-01  -9.1712284088e-01  -9.0183424950e-01   
9   8.8619679213e-02  -4.4551330805e-01  -1.2711701393e+00  -1.2401897907e+00   

       chroma_cens.7      chroma_cens.8  ...        tonnetz.39  \
0           kurtosis           kurtosis  ...               std   
1                 08                 09  ...                04   
2                NaN                NaN  ...               NaN   
3   2.6914546490e+00   8.6686819792e-01  ...  5.4125156254e-02   
4   1.0540937185e+00   1.0810308903e-01  ...  6.3831120729e-02   
5  -9.2723226547e-01   6.6661673784e-01  ...  4.0730185807e-02   
6   1.6045658588e+00   5.2121698856e-01  ...  7.4357867241e-02   
7  -8.1641042233e-01   4.3850939721e-02  ...  9.5002755523e-02   
8  -6.6844828427e-02  -2.9103723168e-01  ...  1.0371652246e-01   
9  -1.3437650204e+00  -9.0560036898e-01  ...  1.4169253409e-01   

         tonnetz.40        tonnetz.41               zcr             zcr.1  \
0               std               std          kurtosis               max   
1                05                06                01                01   
2               NaN               NaN               NaN               NaN   
3  1.2225749902e-02  1.2110591866e-02  5.7588901520e+00  4.5947265625e-01   
4  1.4211839065e-02  1.7740072682e-02  2.8246941566e+00  4.6630859375e-01   
5  1.2690781616e-02  1.4759079553e-02  6.8084154129e+00  3.7500000000e-01   
6  1.7951935530e-02  1.3921394013e-02  2.1434211731e+01  4.5214843750e-01   
7  2.2492416203e-02  2.1355332807e-02  1.6669036865e+01  4.6972656250e-01   
8  2.5541320443e-02  2.3846302181e-02  4.1645809174e+01  2.5048828125e-01   
9  2.0426128060e-02  2.5417611003e-02  8.1665945053e+00  5.4687500000e-01   

              zcr.2             zcr.3             zcr.4             zcr.5  \
0              mean            median               min              skew   
1                01                01                01                01   
2               NaN               NaN               NaN               NaN   
3  8.5629448295e-02  7.1289062500e-02  0.0000000000e+00  2.0898721218e+00   
4  8.4578499198e-02  6.3964843750e-02  0.0000000000e+00  1.7167237997e+00   
5  5.3114086390e-02  4.1503906250e-02  0.0000000000e+00  2.1933031082e+00   
6  7.7514506876e-02  7.1777343750e-02  0.0000000000e+00  3.5423245430e+00   
7  4.7224905342e-02  4.0039062500e-02  9.7656250000e-04  3.1898307800e+00   
8  1.8387714401e-02  1.56250000

In [26]:
len(feats.columns)

519

# Features
- I'm not sure what the first two columns contain. 
- There are **519 attributes**
  - I don't understand what any of them are yet

I want to join `feats` and `track_attr`, but have to change feats to remove the first two rows, while saving the columns?

Maybe it's easier to just add the urls based on the `track_id`.. let's try that

In [55]:
feats2 = feats
feats2.iloc[3:, :] # all cols, rows 3-end

feature        chroma_cens      chroma_cens.1      chroma_cens.2  \
3            2   7.1806526184e+00   5.2303090096e+00   2.4932080507e-01   
4            3   1.8889633417e+00   7.6053929329e-01   3.4529656172e-01   
5            5   5.2756297588e-01  -7.7654317021e-02  -2.7961030602e-01   
6           10   3.7022454739e+00  -2.9119303823e-01   2.1967420578e+00   
7           20  -1.9383698702e-01  -1.9852678478e-01   2.0154602826e-01   
8           26  -6.9953453541e-01  -6.8415790796e-01   4.8824872822e-02   
9           30  -7.2148716450e-01  -8.4855991602e-01   8.9090377092e-01   
10          46  -1.1970755458e-01  -8.5881441832e-01   2.3625464439e+00   
11          48  -1.0540534258e+00   9.3233942986e-01   5.2806353569e-01   
12         134   9.1844475269e-01   6.7414724827e-01   5.7781827450e-01   
13         135   9.7840219736e-01   6.2357187271e-01   1.1292605400e+00   
14         136   9.1500061750e-01  -6.4347624779e-01  -4.6050721407e-01   
15         137   1.6071908474e+00   1.1996492147e+00   7.4576503038e-01   
16         138   1.5014034510e+00   9.5751410723e-01   1.2766846418e+00   
17         139  -2.0868886262e-02   4.3232953548e-01   3.3127766848e-01   
18         140   5.3357946873e-01  -6.2388461828e-01  -1.0862048864e+00   
19         141   1.7289805412e-01  -2.8480431437e-01  -1.1696622372e+00   
20         142  -5.8126991987e-01   3.1994838715e+00   1.2983456850e+00   
21         144   1.2120676786e-01  -2.5307545438e-02   5.4132652283e-01   
22         145   9.5405399799e-01   3.1532418728e-01   1.5383350663e-02   
23         146  -7.6371264458e-01  -3.4210473299e-01  -2.7396461368e-01   
24         147   1.1157943867e-02   6.9270282984e-02   1.5522817373e+00   
25         148  -4.5395046473e-01  -9.2226713896e-01  -8.4576714039e-01   
26         149  -3.8673317432e-01  -5.4559791088e-01  -6.3864940405e-01   
27         150  -8.3042584360e-02   1.4213098586e-01  -4.6670320630e-01   
28         151   2.6796996593e+00   1.5645813942e+00   1.6287657022e+00   
29         152  -5.1045733690e-01   1.6490739584e+00   3.4084680676e-01   
30         153   2.4377791882e+00   8.8621097803e-01   2.8999738693e+00   
31         154   4.0440719604e+01  -2.5125887990e-01   6.8547183275e-01   
32         155   1.6119478643e-01   1.5808784962e+00  -2.0023119450e-01   
...        ...                ...                ...                ...   
106547  155290           -1.25692           -1.19482           -1.35312   
106548  155291           -1.54384           -1.32084           -1.19966   
106549  155292           -0.95519          -0.879837           0.480942   
106550  155293           0.116289          -0.728864          -0.404994   
106551  155294           -1.05293          -0.875463           0.108331   
106552  155295           -1.85517           -1.84308            6.12109   
106553  155296           -1.18177          -0.953902          -0.767223   
106554  155297        -0.00766765           0.104604          -0.511487   
106555  155298          -0.956019           -0.96121          -0.447053   
106556  155299          -0.378812          -0.692087          -0.869693   
106557  155300          -0.590855          -0.228609          -0.592955   
106558  155301          -0.749575          -0.742979          -0.218289   
106559  155302           -0.73258          -0.203825          -0.883129   
106560  155303           -0.84672           -1.16109          -0.724193   
106561  155304          -0.420769         -0.0992737          -0.251443   
106562  155305          -0.832872           -0.80987          -0.601314   
106563  155306           0.644815          -0.879404           -1.14923   
106564  155307          -0.446937              1.235          -0.250855   
106565  155308           0.647675            1.05856            1.65495   
106566  155309           0.189282            1.26453          -0.444038   
106567  155310           0.124136           0.958202          0.0255369   
106568  155311           -1.0273

Ok, `feature` in `feat2` dataframe now maps to `track_attr["track_id"]`. Let's try to use a merge to add a `url` column!

In [56]:
feats3 = pd.merge(feats2, track_attr, how='inner', left_on = 'feature', right_on = 'track_id')


In [57]:
feats3.head(5)

feature chroma_cens chroma_cens.1 chroma_cens.2 chroma_cens.3 chroma_cens.4  \
0    1296      2.0335       1.22087     -0.434886      0.377794      0.510127   
1    1297   -0.538041     -0.848343     -0.353566     -0.866975      -1.55719   
2    1298    0.120188      0.759824       0.90748      0.687974      -0.25387   
3    1299   -0.498199      0.489884       1.31216     -0.879331     -0.741421   
4    1300     16.9894       43.4096       41.3554     0.0174829       164.881   

  chroma_cens.5 chroma_cens.6 chroma_cens.7 chroma_cens.8  ...  \
0     -0.215082       2.60096      0.893777     0.0236751  ...   
1      -1.46657     -0.146988        1.2515      0.252045  ...   
2    -0.0629012      0.454025     -0.446084     -0.367723  ...   
3     -0.703711     -0.659329     -0.313566      0.127728  ...   
4       194.719       73.1801       439.131       188.034  ...   

  track_instrumental track_interest track_language_code track_listens  \
0                  0            396                  en            62   
1                  0            978                  en           107   
2                  0            809                  en           340   
3                  0            513                  en           130   
4                  0            526                  en           111   

  track_lyricist track_number track_publisher       track_title  \
0            NaN            9             NaN  The Picket Fence   
1            NaN           10             NaN       White Walls   
2            NaN           11             NaN   Weekend Warrior   
3            NaN           12             NaN      Its Not Like   
4            NaN           13             NaN      Doctor Spins   

                                           track_url urls  
0  http://freemusicarchive.org/music/Newagehillbi...    8  
1  http://freemusicarchive.org/music/Newagehillbi...    8  
2  http://freemusicarchive.org/music/Newagehillbi...    8  
3  http://freemusicarchive.org/music/Newagehillbi...    8  
4  http://freemusicarchive.org/music/Newagehillbi...    8  

[5 rows x 560 columns]

That's not right :-)